File: example/bench/parameters.cpp
#include <exception>
#include <fstream>
#include <iostream>
#include <string>

#include <aux/json_params.hpp>

#include "parameters.hpp"

double bench_params::expected_advance_time() const {
    return cell.realtime_ratio * duration*1e-3 * num_cells;
}
unsigned bench_params::expected_spikes() const {
    return num_cells * duration*1e-3 * cell.spike_freq_hz;
}
unsigned bench_params::expected_spikes_per_interval() const {
    return num_cells * network.min_delay*1e-3/2 * cell.spike_freq_hz;
}
unsigned bench_params::expected_events() const {
    return expected_spikes() * network.fan_in;
}
unsigned bench_params::expected_events_per_interval() const {
    return expected_spikes_per_interval() * network.fan_in;
}

std::ostream& operator<<(std::ostream& o, const bench_params& p) {
    o << "benchmark parameters:\n"
      << "  name:          " << p.name << "\n"
      << "  num cells:     " << p.num_cells << "\n"
      << "  duration:      " << p.duration << " ms\n"
      << "  fan in:        " << p.network.fan_in << " connections/cell\n"
      << "  min delay:     " << p.network.min_delay << " ms\n"
      << "  spike freq:    " << p.cell.spike_freq_hz << " Hz\n"
      << "  cell overhead: " << p.cell.realtime_ratio << " ms to advance 1 ms\n";
    o << "expected:\n"
      << "  cell advance: " << p.expected_advance_time() << " s\n"
      << "  spikes:       " << p.expected_spikes() << "\n"
      << "  events:       " << p.expected_events() << "\n"
      << "  spikes:       " << p.expected_spikes_per_interval() << " per interval\n"
      << "  events:       " << p.expected_events_per_interval()/p.num_cells << " per cell per interval";
    return o;
}

bench_params read_options(int argc, char** argv) {
    using aux::param_from_json;

    bench_params params;
    if (argc<2) {
        std::cout << "Using default parameters.\n";
        return params;
    }
    if (argc>2) {
        throw std::runtime_error("More than command line one option not permitted.");
    }

    std::string fname = argv[1];
    std::cout << "Loading parameters from file: " << fname << "\n";
    std::ifstream f(fname);

    if (!f.good()) {
        throw std::runtime_error("Unable to open input parameter file: "+fname);
    }

    nlohmann::json json;
    json << f;

    param_from_json(params.name, "name", json);
    param_from_json(params.num_cells, "num-cells", json);
    param_from_json(params.duration, "duration", json);
    param_from_json(params.network.min_delay, "min-delay", json);
    param_from_json(params.network.fan_in, "fan-in", json);
    param_from_json(params.cell.realtime_ratio, "realtime-ratio", json);
    param_from_json(params.cell.spike_freq_hz, "spike-frequency", json);

    for (auto it=json.begin(); it!=json.end(); ++it) {
        std::cout << "  Warning: unused input parameter: \"" << it.key() << "\"\n";
    }
    std::cout << "\n";

    return params;
}

File: validation/ref/neuron/gap_junction/parameters.py
import json

def from_json(o, key):
    if key in o:
        return o[key]
    else:
        raise Exception(str('parameter "'+ key+ '" not in input file'))

class cell_parameters:
    def __repr__(self):
        s = "cell parameters\n" \
            "  depth        : {0:10d}\n" \
            "  branch prob  :     [{1:5.2f} : {2:5.2f}]\n" \
            "  compartments :     [{3:5d} : {4:5d}]\n" \
            "  lengths      :     [{5:5.1f} : {6:5.1f}]\n" \
            .format(self.max_depth,
                    self.branch_probs[0], self.branch_probs[1],
                    self.compartments[0], self.compartments[1],
                    self.lengths[0], self.lengths[1])
        return s

    def __init__(self, data=None):
        if data:
            self.max_depth    = from_json(data, 'depth')
            self.branch_probs = from_json(data, 'branch-probs')
            self.compartments = from_json(data, 'compartments')
            self.lengths      = from_json(data, 'lengths')
            self.synapses     = from_json(data, 'synapses')
        else:
            self.max_depth    = 5
            self.branch_probs = [1.0, 0.5]
            self.compartments = [20, 2]
            self.lengths      = [200, 20]
            self.synapses     = 1

class model_parameters:
    def __repr__(self):
        s = "parameters\n" \
            "  name         : {0:>10s}\n" \
            "  cells        : {1:10d}\n" \
            "  duration     : {2:10.0f} ms\n" \
            "  min delay    : {3:10.0f} ms\n" \
            .format(self.name, self.num_cells, self.duration, self.min_delay)
        s+= str(self.cell)
        return s

    def __init__(self, filename=None):
        if filename:
            with open(filename) as f:
                data = json.load(f)
                self.name      = from_json(data, 'name')
                self.num_cells = from_json(data, 'num-cells')
                self.duration  = from_json(data, 'duration')
                self.min_delay = from_json(data, 'min-delay')
                self.cell      = cell_parameters(data)

        else:
            self.name      = 'default'
            self.num_cells = 2
            self.duration  = 100
            self.min_delay = 10
            self.cell = cell_parameters()
